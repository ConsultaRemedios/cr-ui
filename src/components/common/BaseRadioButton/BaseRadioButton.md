### Basic Usage
 ### Checked
```js
<BaseRadioButton name="gender" value="male" :checked="true" label="Checkbox"/>
```
### Unchecked
```js
<BaseRadioButton name="gender" value="male" :checked="false" label="Checkbox"/>
```

### Update checked prop using the one way data flow

```js
new Vue({
  data(){
    return { gender: '' };
  },
  methods: {
    onChange(payload) {
      this.gender = payload.value
    }
  },
  template: `
    <div>
      <BaseRadioButton name="gender" value="male" :checked="gender == 'male'" label="Male" @change="onChange"/>
      <br>
      <BaseRadioButton name="gender" value="female" :checked="gender == 'female'" label="Female" @change="onChange"/>
    </div>
  `
})
```