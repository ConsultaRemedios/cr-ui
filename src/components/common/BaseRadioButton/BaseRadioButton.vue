<template>
  <div>
    <label :class="[$style.wrapper, {[$style.wrapperSelected]: checked}]">
      <input
        type="radio"
        :name="name"
        :value="value"
        @change="onChange"
        :checked="checked"
        v-bind="$attrs"
      />
      <span
        :class="$style.checkmark"
      ></span>
      <span v-if="label" :class="$style.label">{{ label }}</span>
    </label>
  </div>
</template>

<script>
  export default {
    inheritAttrs: false,
    name: 'BaseRadioButton',

    props: {
      label: {
        type: String,
        default: '',
      },
      name: {
        type: String,
        required: true,
      },

      value: {
        type: String,
        required: true,
      },

      checked: {
        type: Boolean,
        default: false,
      },

      model: {
        type: String,
        default: '',
      },
    },

    methods: {
      onChange() {
        this.$emit('change', {
          name: this.model || this.name,
          value: this.value,
        });
      },
    },
  };
</script>

<style module>
  .wrapper {
    cursor: pointer;
    display: block;
    position: relative;
    user-select: none;
    cursor: pointer;
    min-height: 25px;
  }

  .wrapper::before {
    content: '';
    display: inline-block;
    width: 25px;
    height: 25px;
    margin-right: 5px;
    vertical-align: top;
    font-weight: 400;
  }

  .wrapper:hover .checkmark {
    border-color: #ccc
  }

  .label {
    display: inline-block;
    vertical-align: top;
    padding-top: 3px;
    font-weight: 400;
    line-height: 1.4;
  }

  .checkmark {
    position: absolute;
    top: 0;
    left: 0;
    height: 25px;
    width: 25px;
    background-color: white;
    border-radius: 50%;
    border: 2px solid #BDBDBD;
  }

  .checkmark:after {
    content: "";
    position: absolute;
    display: none;
    top: 8px;
    left: 8px;
    width: 9px;
    height: 9px;
    border-radius: 50%;
    background: white;
  }

  .wrapper input {
    position: absolute;
    opacity: 0;
  }

  .wrapperSelected .checkmark {
    background-color: #00AAE5;
    display: block;
    border: 0;
  }

  .wrapperSelected .checkmark:after {
    display: block;
  }
</style>
