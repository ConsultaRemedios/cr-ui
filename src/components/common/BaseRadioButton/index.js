import BaseRadioButton from './BaseRadioButton.vue';

export default BaseRadioButton;
