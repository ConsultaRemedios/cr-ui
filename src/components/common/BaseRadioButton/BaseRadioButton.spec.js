import { mount } from "@vue/test-utils";
import BaseRadioButton from './BaseRadioButton.vue';

describe('BaseRadioButton', () => {
  const defaultProps = {
    name: 'gender',
    value: 'male',
  }
  const mountBaseRadioButton = (props = {}, attributes = {} ) => {
    const propsData = { ...defaultProps, ...props };
    const attrs = { ...attributes };
    return mount(BaseRadioButton, { propsData, attrs });
  };

  describe('Defaults', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = mountBaseRadioButton();
    });

    it('assigns checked prop as false', () => {
      expect(wrapper.vm.checked).toBe(false);
    });

    it('assigns label prop as an empty string', () => {
      expect(wrapper.vm.label).toBe('');
    });

    it('assigns label prop as no empty string', () => {
      wrapper = mountBaseRadioButton({ label: 'Male' });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.vm.label).toBe('Male');
    });
  });

  describe('Behavior', () => {
    describe('When it is checked and a click happens', () => {
      let wrapper;

      beforeAll(() => {
        wrapper = mountBaseRadioButton({ checked: true });
        wrapper.find('input').trigger('click');
      });

      it('does not emmit a change event', () => {
        const event = wrapper.emitted();
        expect(event).toMatchObject({});
      });
    })

    describe('when it is not checked and a click happens', () => {
      describe('and model prop is empty', () => {
        it('emmits a change event with proper params', () => {
          const wrapper = mountBaseRadioButton({ checked: false });
          wrapper.find('input').trigger('change');
          const event = wrapper.emitted().change[0][0];

          expect(event).toMatchObject({ name: 'gender', value: 'male' });
        });
      });

      describe('and model prop is not empty', () => {
        it('emmits a change event with proper params', () => {
          const wrapper = mountBaseRadioButton({ checked: false, model: 'customModel' });
          wrapper.find('input').trigger('change');
          const event = wrapper.emitted().change[0][0];

          expect(event).toMatchObject({ name: 'customModel', value: 'male' });
        });
      });
    });
  });

  describe('Render', () => {
    describe('when it is checked', () => {
      it('matches the snapshot', () => {
        const wrapper = mountBaseRadioButton({ checked: true });
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('when it is not checked', () => {
      it('matches the snapshot', () => {
        const wrapper = mountBaseRadioButton({ checked: false });
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('When there is label', () => {
      it('renders the snapshot', () => {
        const wrapper = mountBaseRadioButton({ label: 'Male' });
        expect(wrapper.element).toMatchSnapshot();
        expect(wrapper.vm.label).toBe('Male');
      });
    });

    describe('When there is no label', () => {
      it('renders the snapshot', () => {
        const wrapper = mountBaseRadioButton();
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('When there is id', () => {
      it('renders the snapshot', () => {
        const wrapper = mountBaseRadioButton(null, { id: 'radio-male' });
        expect(wrapper.element).toMatchSnapshot();
        expect(wrapper.vm.$attrs.id).toBe('radio-male');
      });
    });

    describe('When there is no id', () => {
      it('renders the snapshot', () => {
        const wrapper = mountBaseRadioButton();
        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });
});
