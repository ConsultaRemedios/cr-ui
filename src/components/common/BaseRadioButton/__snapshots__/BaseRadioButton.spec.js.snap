// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseRadioButton Defaults assigns label prop as no empty string 1`] = `
<div>
  <label
    class="wrapper"
  >
    <input
      name="gender"
      type="radio"
      value="male"
    />
     
    <span
      class="checkmark"
    />
     
    <span
      class="label"
    >
      Male
    </span>
  </label>
</div>
`;

exports[`BaseRadioButton Render When there is id renders the snapshot 1`] = `
<div>
  <label
    class="wrapper"
  >
    <input
      id="radio-male"
      name="gender"
      type="radio"
      value="male"
    />
     
    <span
      class="checkmark"
    />
     
    <!---->
  </label>
</div>
`;

exports[`BaseRadioButton Render When there is label renders the snapshot 1`] = `
<div>
  <label
    class="wrapper"
  >
    <input
      name="gender"
      type="radio"
      value="male"
    />
     
    <span
      class="checkmark"
    />
     
    <span
      class="label"
    >
      Male
    </span>
  </label>
</div>
`;

exports[`BaseRadioButton Render When there is no id renders the snapshot 1`] = `
<div>
  <label
    class="wrapper"
  >
    <input
      name="gender"
      type="radio"
      value="male"
    />
     
    <span
      class="checkmark"
    />
     
    <!---->
  </label>
</div>
`;

exports[`BaseRadioButton Render When there is no label renders the snapshot 1`] = `
<div>
  <label
    class="wrapper"
  >
    <input
      name="gender"
      type="radio"
      value="male"
    />
     
    <span
      class="checkmark"
    />
     
    <!---->
  </label>
</div>
`;

exports[`BaseRadioButton Render when it is checked matches the snapshot 1`] = `
<div>
  <label
    class="wrapper wrapperSelected"
  >
    <input
      name="gender"
      type="radio"
      value="male"
    />
     
    <span
      class="checkmark"
    />
     
    <!---->
  </label>
</div>
`;

exports[`BaseRadioButton Render when it is not checked matches the snapshot 1`] = `
<div>
  <label
    class="wrapper"
  >
    <input
      name="gender"
      type="radio"
      value="male"
    />
     
    <span
      class="checkmark"
    />
     
    <!---->
  </label>
</div>
`;
